admin/BundleSearchIndexesDump

```diff
@@ -0,0 +1,210 @@
+
+use DBDefs;
+use MusicBrainz::Script::MBDump;
+
+################################################################################
+
+sub _debug { print "[debug] ", (sprintf shift, @_), "\n" if $debug_flag; }
+sub _info { print "[info] ", (sprintf shift, @_), "\n"; }
+sub _warn { warn "[warning] ", (sprintf shift, @_), "\n"; }
+sub _error { warn "[error] ", (sprintf shift, @_), "\n"; }
+
+my $BACKUP_STAMP = $ENV{BACKUP_STAMP};
+my $COMPRESSION_LEVEL = DBDefs->SEARCH_INDEXES_DUMP_COMPRESSION_LEVEL
+
+my @backup_dirs;
+my @extraneous_files;
+
+opendir(my $wdh, $working_dir) or die $!;
+for my $filename (sort grep { ! /^\.\.?$/ } readdir $wdh)
+{
+    if ($filename =~ /^backup_${BACKUP_STAMP}_/)
+    {
+        push @backup_dirs, $filename;
+    }
+    else
+    {
+        push @extraneous_files, $filename;
+    }
+}
+closedir $wdh;
+
+die "$working_dir has extraneous files: @extraneous_files\n"
+    if scalar @extraneous_files;
+
+my @public_domain_collections = qw(
+    area
+    artist
+    event
+    instrument
+    label
+    place
+    recording
+    release
+    release-group
+    series
+    url
+    work
+);
+
+for my $backup_dir (@backup_dirs)
+{
+    my $collection = $backup_dir =~ s/^backup_\d{8}-\d{6}_//r;
+    my $dump_dir = $collection;
+    opendir(my $bdh, "$working_dir/$backup_dir") or die $!;
+    for my $filename (sort grep { ! /^\.\.?$/ } readdir $bdh)
+    {
+        push @extraneous_files, $filename
+            if ! $filename =~ "backup.properties|snapshot.shard1";
+    }
+    closedir $bdh;
+    die "$working_dir/$backup_dir has extraneous files: @extraneous_files\n"
+        if scalar @extraneous_files;
+
+    my $working_dump_dir = "$working_dir/$dump_dir";
+    move("$working_dir/$backup_dir", $working_dump_dir) or die $!;
+    move("$working_dump_dir/snapshot.shard1", "$working_dump_dir/index") or die $!;
+
+    my $license_file = grep { /^$collection$/ } @public_domain_collections
+        ? "$FindBin::Bin/COPYING-PublicDomain"
+        : "$FindBin::Bin/COPYING-CCShareAlike";
+    copy($license_file, "$working_dump_dir/COPYING") or die $!;
+
+    my $readme_text = <<EOF;
+The files in this directory are a snapshot of the search index for
+'$collection' in the MusicBrainz database, in a format suitable for
+use with an Apache Solr server.  To import them, you need a compatible
+version of the MusicBrainz simple Solr search server schema (mbsssss).
+EOF
+    _write_file($working_dump_dir, 'README', $readme_text);
+
+    my $tar_file="$collection.tar.xz"
+    log("Creating $tar_file");
+    system
+        'bash', ' -c',
+            'set -o pipefail; ' .
+            'tar' .
+            ' -C ' . shell_quote($working_dir) .
+            ' --create' .
+            ' --group=solr:8983' .
+            ' --owner=solr:8983' .
+            ' --remove-files' .
+            ' --verbose' .
+            ' -- ' .
+            shell_quote($dump_dir) .
+            " | xz --threads=0 -$COMPRESSION_LEVEL" .
```

Comment:
Not really, I choosed `xz` for convenience: because it is already installed with MusicBrainz Server (to handle `json` and `sample` dumps) and `xzcat` is already installed with Solr (Docker Official image). And also because it is very versatile: its fastest compression level is faster than `bzip2` without sacrificing as much as `gzip` and the curve up to default compression level allows to tune it significantly if needed.

If I was to pick a new algorithm up, I would probably look at Zstandard. It seems to be very efficient when it comes to compress a lot of small files such as search indexes.

Reply:
Cool. Comparing xz with zstd or lz4 here would be very interesting indeed.

Reply:
So I tested these 3 compression algorithms with real data (work search index) and Zstandard is by far the best compromise:

| algorithm / level | time | space savings |
| -- | -- | -- |
| lz4 / 1 | 3.65 | 89.93% |
| lz4 / 2 | 3.60 | 89.93% |
| lz4 / 3 | 56.34 | 76.54% |
| lz4 / 4 | 63.94 | 76.04% |
| lz4 / 5 | 68.01 | 75.88% |
| lz4 / 6 | 71.25 | 75.81% |
| xz / 0 | 307.24 | 62.39% |
| xz / 1 | 493.05 | 60.14% |
| xz / 2 | 795.86 | 59.06% |
| xz / 3 | 1299.95 | 58.03% |
| zstd / 1 | 11.40 | 66.59% |
| zstd / 2 | 19.64 | 64.87% |
| zstd / 3 | 31.75 | 63.90% |
| zstd / 4 | 34.64 | 63.41% |
| zstd / 5 | 69.01 | 62.43% |
| zstd / 6 | 92.93 | 61.47% |

Changed to use `zstd` in https://github.com/metabrainz/musicbrainz-server/pull/1407/commits/bfcaa355b5216f3fa5abb32e4208de4e9d4c41be (and https://github.com/metabrainz/musicbrainz-server/pull/1407/commits/15631dbb768e9cf10d70d7b73b125a83ef07c03f)

Reply:
Nice! Glad you found time to measure that, sounds like it should make the process much more efficient.